src/java.base/share/classes/sun/security/pkcs12/MacData.java

```diff
@@ -107,29 +142,161 @@ class MacData {
             this.digest = digest.clone();
         }
 
-        this.macSalt = salt;
-        this.iterations = iterations;
+        if (!(params instanceof PBEParameterSpec p)) {
+            throw new IllegalArgumentException("unsupported parameter spec");
+        }
+
+        if (this.pbmac1Keystore) {
+            this.macSalt = p.getSalt();
+            this.iterations = p.getIterationCount();
+            this.kdfHmac = kdfHmac;
+            this.hmac = hmac;
+            this.keyLength = keyLength;
+        } else {
+            this.macSalt = p.getSalt();
+            this.iterations = p.getIterationCount();
+            this.kdfHmac = null;
+            this.hmac = null;
+            this.keyLength = 0;
+        }
 
         // delay the generation of ASN.1 encoding until
         // getEncoded() is called
         this.encoded = null;
+    }
 
+    /*
+     * Destroy the key obtained from getPBEKey().
+     */
+    static void destroyPBEKey(SecretKey key) {
+        try {
+            key.destroy();
+        } catch (DestroyFailedException e) {
+            // Accept this
+        }
     }
 
-    String getDigestAlgName() {
-        return digestAlgorithmName;
+    static Mac getMac(String macAlgorithm, char[] password,
+            PBEParameterSpec params, byte[] data,
+            String kdfHmac, String hmac)
+            throws NoSuchAlgorithmException, InvalidKeySpecException,
+            InvalidKeyException, InvalidAlgorithmParameterException {
+        SecretKeyFactory skf;
+        SecretKey pbeKey;
+        Mac m;
+
+        if (macAlgorithm.startsWith("PBEWith")) {
+            m = Mac.getInstance(hmac);
+            int keyLength = m.getMacLength()*8;
+            skf = SecretKeyFactory.getInstance("PBKDF2With" +kdfHmac);
+            pbeKey = skf.generateSecret(new PBEKeySpec(password,
+                    params.getSalt(), params.getIterationCount(), keyLength));
+        } else {
+            hmac = macAlgorithm;
+            m = Mac.getInstance(hmac);
+            PBEKeySpec keySpec = new PBEKeySpec(password);
+            skf = SecretKeyFactory.getInstance("PBE");
+            pbeKey = skf.generateSecret(keySpec);
+        }
+
+        try {
+            if (macAlgorithm.startsWith("PBEWith")) {
+                 m.init(pbeKey);
+            } else {
+                 m.init(pbeKey, params);
+            }
+        } finally {
+            destroyPBEKey(pbeKey);
```

Comment:
The `PBEKeySpec` object created in both cases should also be cleaned up by calling `keySpec.clearPassword()`.

Reply:
The password in `PBKDF2KeyImpl` is not cleaned. It does have a `clear` method but no `destroy`. Consider adding a `destroy` to it.